src/output_module.h

```diff
@@ -25,28 +25,71 @@
 #ifndef _OUTPUT_MODULE_H
 #define _OUTPUT_MODULE_H
 
+#include <set>
+#include <string>
+#include <vector>
+
 #include "output.h"
 
-struct output_module {
-  const char *shortname;
-  const char *description;
-  int (*add_options)(GOptionContext *ctx);
-
-  // Commands.
-  int (*init)(void);
-  void (*set_uri)(const char *uri, output_update_meta_cb_t meta_info);
-  void (*set_next_uri)(const char *uri);
-  int (*play)(output_transition_cb_t transition_callback);
-  int (*stop)(void);
-  int (*pause)(void);
-  int (*seek)(gint64 position_nanos);
-
-  // parameters
-  int (*get_position)(gint64 *track_duration, gint64 *track_pos);
-  int (*get_volume)(float *);
-  int (*set_volume)(float);
-  int (*get_mute)(int *);
-  int (*set_mute)(int);
+class OutputModule {
+ public:
+  struct Options {
+    virtual std::vector<GOptionGroup*> GetOptionGroups(void) = 0;
+  };
+
+  typedef struct TrackState {
+    int64_t duration_ns;
+    int64_t position_ns;
+  } TrackState;
+
+  typedef enum Result { kSuccess = 0, kError = -1 } Result;
+
+  OutputModule(Output::PlaybackCallback play = nullptr,
```

Comment:
Currently we always use them, but they are not required per say. The output will continue to operate even if it can't notify. I was just writing it to be flexible. Whether that's necessary or not? I dunno.

Reply:
Let's add flexibility when we need it, not earlier.